rfd/0173-terraform-machine-id.md

```diff
@@ -0,0 +1,326 @@
+The `tbot` instance will not persist its certificates on the disk.
+
+A typical GitHub actions provider configuration would look like:
+
+```hcl
+provider "teleport" {
+  addr               = "mytenant.teleport.sh:443"
+  joining = {
+    token  = "gha-runner"
+    method = "github"
+  }
+}
+```
+
+As Teleport does not support yet reusable bot tokens so the `token` join method will not be supported.
+_Note: this might change in the future with the `BotInstance` resource decoupling the bot from the
+secret token._
+
+Implementation-wise, this would reuse the
+same [in-process tbot library](https://github.com/gravitational/teleport/tree/rfd/173-terraform-machine-id/integrations/operator/embeddedtbot)
+we used for the Teleport operator.
+
+##### Note regarding CGO
+
+The Terraform provider is currently built without CGO but `tbot` depends on `lib/auth`, which depends on `lib/services`
+which depends on CGO. As CGO is not supported in Hashicorp Terraform Cloud, the embedded tbot
+will rely on the work currently done by @rosstimothy and the MachineID team to remove the CGO dependency from tbot.
+
+#### Resource bootstrapping
+
+We will automatically create the resources required by the Terraform provider before executing Terraform:
+- the `terraform-provider` role (this is an upsert so we can add new rules after a provider update). This resource does
+  not expire.
+- the `terraform-provider-<user>-<random hash>` bot allowed to issue certificates for the `terraform-provider` role.
+  This resource expires.
+- the `terraform-provider-<user>-<random hash>` secret random provision token allowing to join as
+  the `terraform-provider-<user>-<random hash>` bot. This resource expires but will be consumed automatically on join.
+
+##### Discarded appraoches
+
+Creating those resources requires passing an MFA challenge, this means being built with libfido2 and CGO.
+This would make the provider unable to run on Hashicorp Cloud Platform, which is requested by many Teleport users.
+
+Two alternatives were considered but discarded:
+
+- Have Terraform re-exec `tctl` or `tsh` to pass the MFA challenge and create the resources. The issue with this
+  approach is that we don't have access to stdin/stdout/stderr to prompt the user for MFA.
+- Have `tsh`/`tctl` create the resources first, then exec into Terraform by passing the token via environment variables.
+  This would force users to invoke `tctl` each time instead of terraform which is cumbersome. Handling arguments
+  properly would also be a challenge.
+
+##### Proposed approach
+
+We will add a `tctl terraform-helper` command which will
+- upsert the terraform role
+- create a temporary bot (1hour) and token
+- run a oneshot tbot to retrieve certificates
+- set certificates in the shell's environment variables using the already supported `TF_TELEPORT_IDENTITY_FILE_BASE64`
+
+The full flow will look like
+
+```
+$ eval $(tctl terraform-local-helper)
```

Comment:
I think this approach is fine but I'm not a fan of the command name. How about something like `tctl terraform env` because it basically outputs the environment variables you need to set in your shell? We have some [similar prior art](https://github.com/gravitational/teleport/blob/1acd9cb4827e5e587114e43f0acfc579102aaa36/tool/tsh/common/tsh.go#L938-L939) too.